File: 2023/go/day12/day12.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

// TODO: FINISH THIS - please look somewhere else
func main() {

	// read the input file
	bytes, err := os.ReadFile("input.txt")
	check(err)
	var count int

	inputs := strings.Split(string(bytes), "\n")
	for i := 0; i < len(inputs); i++ {
		input := strings.Split(inputs[i], " ")
		springs := input[0]
		groups := strings.Split(input[1], ",")

		groupsInt := make([]int, len(groups))
		for idx, i := range groups {
			j, err := strconv.Atoi(i)
			if err != nil {
				panic(err)
			}
			groupsInt[idx] = j
		}

		var springB strings.Builder
		for i := 0; i < 5; i++ {
			if i == 0 {
				springB.WriteString(springs)
			} else {
				springB.WriteString("?" + springs)
			}
		}

		var groupB []int
		for i := 0; i < 5; i++ {
			groupB = append(groupB, groupsInt...)
		}

		count += countArrangements(springB.String(), groupB)
		// fmt.Println("---springs: ", springB.String())
		// fmt.Println("---groups: ", groupB)
	}
	// a := []int{3, 2, 1}

	fmt.Print("--count", count)
}

func countArrangements(spring string, group []int) int {

	springArr := strings.Split(spring, "")

	if len(spring) == 0 {
		if len(group) == 0 {
			// fmt.Println("--")
			return 1
		}
		return 0
	}

	if len(group) == 0 {
		if strings.Contains(spring, "#") {
			return 0
		}
		// fmt.Println("???")
		return 1
	}

	sum := 0
	if springArr[0] == "." || springArr[0] == "?" {
		sum += countArrangements(spring[1:], group)
	}

	if springArr[0] == "#" || springArr[0] == "?" {

		if group[0] <= len(spring) &&
			!strings.Contains(spring[0:group[0]], ".") && (group[0] == len(spring) || springArr[group[0]] != "#") {

			if group[0] == len(spring) {
				sum += countArrangements("", group[1:])
			} else {
				sum += countArrangements(spring[(group[0]+1):], group[1:])
			}
		}

	}

	// fmt.Println("---SUM", sum)

	return sum
}

File: 2023/ts/day12/day12.ts
// TODO: FINISH THIS - please look somewhere else
export function partOne(data: string) {
  let sum: number = 0;
  const lines = data.split("\n");

  lines.forEach((spring) => {
    const springValue = spring.split(" ")[0];
    const springNumbers = spring
      .split(" ")[1]
      .split(",")
      .map((e) => Number(e));
    // console.log("--spring", springValue, springNumbers);
    sum += countSprings(springValue, springNumbers);
  });

  return sum;
}

export function partTwo(data: string) {
  let sum: number = 0;
  const lines = data.split("\n");

  lines.forEach((spring) => {
    const springValue = spring.split(" ")[0];
    const springNumbers = spring
      .split(" ")[1]
      .split(",")
      .map((e) => Number(e));

    let newSpringValue = "";
    let newSpringNumber: number[] = [];

    let i = 1;
    while (i < 6) {
      newSpringValue = springValue + "?" + newSpringValue;
      newSpringNumber.push(springNumbers);
      i++;
    }

    // console.log(
    //   "--spring",
    //   newSpringValue.substring(0, newSpringValue.length - 1),
    //   newSpringNumber.flat()
    // );

    sum += countSprings(
      newSpringValue.substring(0, newSpringValue.length - 1),
      newSpringNumber.flat()
    );
  });

  return sum;
}

export function countSprings(spring: string, group: number[]): number {
  // console.log(group, spring);

  if (spring.length === 0) {
    if (group.length === 0) {
      return 1;
    }
    return 0;
  }

  if (group.length === 0) {
    if (spring.includes("#")) {
      return 0;
    }
    return 1;
  }

  let sum = 0;

  if (spring[0] === "." || spring[0] === "?") {
    sum += countSprings(spring.substring(1), group);
    // console.log("---start", spring, group, sum);
  }

  if (spring[0] === "#" || spring[0] === "?") {
    if (
      group[0] <= spring.length &&
      !spring.substring(0, group[0]).includes(".") &&
      (group[0] === spring.length || spring[group[0]]) != "#"
    ) {
      // console.log("-------", spring, group[0], "---", group, "sum: ", sum);

      if (group[0] === spring.length) {
        sum += countSprings("", group.slice(1));
        // console.log("---if", spring, group, sum);
      } else {
        sum += countSprings(spring.substring(group[0] + 1), group.slice(1));
        // console.log("sum: ", sum);
        // console.log("---else", spring, group, sum);
      }
    }
  }

  // console.log("=============");

  return sum;
}
